File: src/main/java/me/glorantq/aboe/chat/client/chat/ModifiedGuiChat.java
package me.glorantq.aboe.chat.client.chat;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiChat;
import net.minecraft.client.gui.GuiPlayerInfo;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Collections;
import java.util.List;

@SideOnly(Side.CLIENT)
public class ModifiedGuiChat extends GuiChat {
    private final Logger logger = LogManager.getLogger("ModifiedGuiChat");

    private boolean drawPlayerTooltip = false;
    private String bestNameMatch = "";
    private String currentPlayerName = "";

    ModifiedGuiChat() {
        super();
    }

    @Override
    @SuppressWarnings("unchecked")
    protected void keyTyped(char p_73869_1_, int p_73869_2_) {
        super.keyTyped(p_73869_1_, p_73869_2_);

        if (p_73869_1_ == '@') {
            bestNameMatch = "";
        }

        String chatText = inputField.getText();

        if (drawPlayerTooltip) {
            if (bestNameMatch.length() > 0 && p_73869_1_ == '\t') {
                inputField.deleteWords(-1);
                inputField.writeText("@" + bestNameMatch);

                logger.info("Successful autocompelte for: {}", bestNameMatch);

                bestNameMatch = "";
                currentPlayerName = "";
                drawPlayerTooltip = false;
                return;
            }

            List<GuiPlayerInfo> onlinePlayers = Minecraft.getMinecraft().thePlayer.sendQueue.playerInfoList; // GuiIngame#L441

            System.out.println(chatText);

            int wordStart = inputField.getNthWordFromCursor(-1);

            int substringStart = wordStart + 1;
            int substringEnd = Math.min(chatText.length(), inputField.getCursorPosition());

            System.out.printf("%d %d %d\n", substringStart, substringEnd, inputField.getCursorPosition());

            if (substringStart > substringEnd) {
                drawPlayerTooltip = false;
            } else {
                currentPlayerName = chatText.substring(substringStart, substringEnd).trim();

                System.out.println(currentPlayerName);

                bestNameMatch = "";
                if (currentPlayerName.length() > 0) {
                    for (GuiPlayerInfo playerInfo : onlinePlayers) {
                        if (currentPlayerName.length() > playerInfo.name.length()) {
                            continue;
                        }

                        String playerNameSubstring = playerInfo.name.substring(0, currentPlayerName.length());

                        if (playerNameSubstring.equalsIgnoreCase(currentPlayerName)) {
                            bestNameMatch = playerInfo.name;
                        }
                    }
                }
            }
        }
    }

    @Override
    public void drawScreen(int p_73863_1_, int p_73863_2_, float p_73863_3_) {
        drawPlayerTooltip =
                inputField.getText().length() > 0 &&
                        inputField.getText().charAt(inputField.getNthWordFromCursor(-1)) == '@' &&
                        inputField.getText().charAt(inputField.getCursorPosition() - 1) != ' ';

        if (drawPlayerTooltip) {
            int cursorX = inputField.xPosition + fontRendererObj.getStringWidth(inputField.getText().substring(0, inputField.getCursorPosition())) - fontRendererObj.getStringWidth("_");
            cursorX = Math.min(cursorX, inputField.xPosition + inputField.width + fontRendererObj.getStringWidth("_"));

            String renderedName;

            if (bestNameMatch.length() > 0) {
                renderedName = "\u00A7f" + bestNameMatch.substring(0, currentPlayerName.length()) + "\u00A78" + bestNameMatch.substring(currentPlayerName.length(), bestNameMatch.length());
            } else {
                renderedName = bestNameMatch;
            }

            func_146283_a(Collections.singletonList(renderedName), cursorX, inputField.yPosition - inputField.height / 2);
        }
        super.drawScreen(p_73863_1_, p_73863_2_, p_73863_3_);
    }
}

File: src/main/kotlin/me/glorantq/aboe/chat/client/chat/ModifiedGuiChat.kt
package me.glorantq.aboe.chat.client.chat

import cpw.mods.fml.relauncher.Side
import cpw.mods.fml.relauncher.SideOnly
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.Gui
import net.minecraft.client.gui.GuiChat
import net.minecraft.client.gui.GuiPlayerInfo
import org.apache.logging.log4j.LogManager

@SideOnly(Side.CLIENT)
class ModifiedGuiChat internal constructor(defaultText: String) : GuiChat(defaultText) {
    private val logger = LogManager.getLogger("ModifiedGuiChat")

    private var drawPlayerTooltip = false
    private var bestNameMatch = ""
    private var currentPlayerName = ""

    override fun keyTyped(p_73869_1_: Char, p_73869_2_: Int) {
        super.keyTyped(p_73869_1_, p_73869_2_)

        if (p_73869_1_ == '@') {
            bestNameMatch = ""
        }

        val chatText = inputField.text

        if (drawPlayerTooltip) {
            if (bestNameMatch.isNotEmpty() && p_73869_1_ == '\t') {
                inputField.deleteWords(-1)
                inputField.writeText("@$bestNameMatch")

                logger.info("Successful autocompelte for: {}", bestNameMatch)

                bestNameMatch = ""
                currentPlayerName = ""
                drawPlayerTooltip = false
                return
            }

            val onlinePlayers: List<GuiPlayerInfo> = Minecraft.getMinecraft().thePlayer.sendQueue.playerInfoList as List<GuiPlayerInfo>

            val wordStart = inputField.getNthWordFromCursor(-1)

            val substringStart = wordStart + 1
            val substringEnd = Math.min(chatText.length, inputField.cursorPosition)

            if (substringStart > substringEnd) {
                drawPlayerTooltip = false
            } else {
                currentPlayerName = chatText.substring(substringStart, substringEnd).trim { it <= ' ' }

                bestNameMatch = ""
                if (currentPlayerName.isNotBlank()) {
                    for (playerInfo in onlinePlayers) {
                        if (currentPlayerName.length > playerInfo.name.length) {
                            continue
                        }

                        val playerNameSubstring = playerInfo.name.substring(0, currentPlayerName.length)

                        if (playerNameSubstring.equals(currentPlayerName, ignoreCase = true)) {
                            bestNameMatch = playerInfo.name
                            break
                        }
                    }

                    if(bestNameMatch.isBlank()) {
                        if("everyone".substring(0, Math.min(currentPlayerName.length, 8)).equals(currentPlayerName, ignoreCase = true)) {
                            bestNameMatch = "everyone"
                        }
                    }
                } else {
                    bestNameMatch = onlinePlayers[0].name
                }
            }
        }
    }

    override fun drawScreen(p_73863_1_: Int, p_73863_2_: Int, p_73863_3_: Float) {
        drawPlayerTooltip =
                inputField.text.isNotEmpty() &&
                inputField.text[inputField.getNthWordFromCursor(-1)] == '@' &&
                inputField.text[inputField.cursorPosition - 1] != ' '

        if (drawPlayerTooltip && bestNameMatch.isNotEmpty() && currentPlayerName.isNotEmpty()) {
            var cursorX = inputField.xPosition + fontRendererObj.getStringWidth(inputField.text.substring(0, inputField.cursorPosition)) - fontRendererObj.getStringWidth("_")
            cursorX = Math.min(cursorX, inputField.xPosition + inputField.width + fontRendererObj.getStringWidth("_"))

            val renderedName: String =
                    if (bestNameMatch.isNotEmpty()) {
                        "\u00A7f" + bestNameMatch.substring(0, currentPlayerName.length) + "\u00A78" + bestNameMatch.substring(currentPlayerName.length, bestNameMatch.length)
                    } else {
                        bestNameMatch
                    }

            val baseY: Int = inputField.yPosition - inputField.height - mc.fontRenderer.FONT_HEIGHT - 5

            Gui.drawRect(cursorX, baseY, cursorX + mc.fontRenderer.getStringWidth(bestNameMatch) + 10, 10 + mc.fontRenderer.FONT_HEIGHT + baseY, 0xDD000000.toInt())
            drawString(mc.fontRenderer, renderedName, cursorX + 5, baseY + 5, 0x000000)
        }

        super.drawScreen(p_73863_1_, p_73863_2_, p_73863_3_)
    }
}
